File: services/frontend/src/services/dataCatalog.js
import axios from 'axios';

const API_BASE = '/api';

export const registerDataset = async (dataset) => {
    const res = await axios.post(`${API_BASE}/data-catalog/datasets`, dataset);
    return res.data;
};

export const listDatasets = async () => {
    const res = await axios.get(`${API_BASE}/data-catalog/datasets`);
    return res.data;
};

export const searchDatasets = async (query) => {
    const res = await axios.get(`${API_BASE}/data-catalog/search?query=${encodeURIComponent(query)}`);
    return res.data;
};

File: services/backend/app/routers/data_catalog.py
from datetime import datetime
from typing import List, Optional

from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel

router = APIRouter()

# In-memory store for demo purposes
DATASETS = []


class DatasetMetadata(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    fields: List[str]
    owner: Optional[str] = None
    created_at: datetime
    tags: Optional[List[str]] = None


@router.post("/data-catalog/datasets", status_code=status.HTTP_201_CREATED)
def register_dataset(dataset: DatasetMetadata):
    if any(d.id == dataset.id for d in DATASETS):
        raise HTTPException(status_code=400, detail="Dataset ID already exists")
    DATASETS.append(dataset)
    return {"message": "Dataset registered", "dataset_id": dataset.id}


@router.get("/data-catalog/datasets", response_model=List[DatasetMetadata])
def list_datasets():
    return DATASETS


@router.get("/data-catalog/search", response_model=List[DatasetMetadata])
def search_datasets(query: str):
    q = query.lower()
    return [d for d in DATASETS if q in d.name.lower() or any(q in f.lower() for f in d.fields)]
